presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/InstallationManager.java

```diff
@@ -122,20 +191,40 @@ public Response stopUsingRpm(StopType stopType)
             LOGGER.error(e.getCause(), e.getMessage());
             return Response.status(INTERNAL_SERVER_ERROR).entity(e.getMessage()).build();
         }
+        catch (IOException e) {
+            LOGGER.error(e, "Failed to get Presto port number");
+            return Response.status(INTERNAL_SERVER_ERROR).entity("Failed to get Presto port number").build();
+        }
         return Response.status(OK).entity("Presto successfully stopped").build();
     }
 
-    public Response restartUsingRpm()
+    private void gracefulStop()
+            throws PrestoManagerException
+    {
+        try {
+            UriBuilder uriBuilder = fromUri(format("http://localhost:%s", getPrestoPort())).path("/v1/info/state");
+            Response response = client.target(uriBuilder.build()).request(TEXT_PLAIN)
+                    .buildPut(entity(GSON.toJson("SHUTTING_DOWN"), APPLICATION_JSON)).invoke();
+            if (response.getStatus() != 200) {
+                throw new PrestoManagerException("Failed to stop presto gracefully");
+            }
+        }
+        catch (ProcessingException e) {
+            LOGGER.warn(e, "Presto is not running");
+        }
+        catch (IOException e) {
+            throw new PrestoManagerException("Failed to get Presto port number.", e);
+        }
+    }
+
+    public Response restart()
```

Comment:
I realized that `restart` should also be asynchronous since it starts the Presto again.